### PR TITLE
General: Validate if current process OpenPype version is requested version

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Brief description
Added periodical checks of OpenPype version.

## Description
Main change is affecting OpenPype tray application. A dialog with information about outdated OpenPype version is shown. If dialog is closed or ignore button is clicked a tray message will popup.

Ftrack event server statuser will show information if is up to date.

## Changes
- moved few functions in lib and added new for checking if version of currently running process is up to date
- tray will check each 5 minutes if is up to date with expected version
    - will show a dialog with ability to restart tray
- restart of tray will unset `OPENPYPE_VERSION` environment for new process
    - this is to trigger update if needed
- ftrack event server statuser has new category OpenPype version with info about current OpenPype version

## Testing notes:
1. Build OpenPype
2. Create a zip with higher version
3. Optional: Remove versions from appdata...
4. Set path to openpype versions to some empty directory
5. Run tray from the build
6. Move the created zip with higher version to openpype versions dir
7. Wait for 5 minutes
8. Popup should be shown
    - you can press Ignore window will close and tray message will show
    - press "Restart and Change" should trigger restart of tray and also update

### Alternative to not wait 5 minutes (code modification needed)
In build go to `~/openpype/tools/tray/pype_tray.py` on line 375 is method `_on_version_action` modify it to
```
def _on_version_action(self):
    self._version_check_timer.timeout.emit()
    ...
```
Now when you click on version action in tray menu it should trigger version validation.

## Notes
- any changes of messages are welcomed
- we could maybe change color of tray icon or add icon to version action (e.g. exclamation mark) or just change color of action text
- color of "Restart and Change" could be maybe changed to better color?
- It can happen that it is not possible to validate if current version is that which should be used. Should we tell user on start?

### Screenshot of dialog showed from Tray
image

### CODE REVIEW

**Comment by mkolar** [comment]
Task linked: [OP-2360 Validate if current OpenPype version is production version](https://app.clickup.com/t/6658547/OP-2360)

**Comment by tokejepsen** [comment]
Is the wait time (5 minutes) configurable in the settings?

File: openpype/tools/project_manager/project_manager/widgets.py [hunk 1]
```diff
@@ -369,7 +369,7 @@ def __init__(self, project_name, parent):
         cancel_btn = QtWidgets.QPushButton("Cancel", self)
         cancel_btn.setToolTip("Cancel deletion of the project")
         confirm_btn = QtWidgets.QPushButton("Permanently Delete Project", self)
-        confirm_btn.setObjectName("DeleteButton")
+        confirm_btn.setObjectName("WarningButton")
```
Comment (BigRoy) [inline openpype/tools/project_manager/project_manager/widgets.py:372]: Is it correct that GitHub is showing these style changes for Project Manager in this PR? Or has there been a merge conflict?

Comment (iLLiCiTiT) [reply openpype/tools/project_manager/project_manager/widgets.py:372]: I've changed the object name because it's using same colors and has same purpose.

**Comment by iLLiCiTiT** [comment]
> Is the wait time (5 minutes) configurable in the settings?

It wasn't, added.

Does make sense to use decimal points for this validation like each 0.5 minutes?

**Comment by tokejepsen** [comment]
> Does make sense to use decimal points for this validation like each 0.5 minutes?

Unless its a bother to implement later, I would stick with whole minutes. Dont think anyone cant wait 1 minute for an update.

**Comment by iLLiCiTiT** [comment]
> I mean, if there's an Update Now button in the Version pop-up window users could always go down that route. Haven't checked myself but is that there? If not, should it be there? I guess just shutting down OP + starting it manually isn't that much effort either.

It's not there, should be there, not goal of this PR. This is more for artists that start OpenPype only when PC is restarted (once in month) so they do not catch new updates at all.

> Do they get the callback again after 5 minutes?

Yes.

> Also, I feel Wrong OpenPype version in the dialog bar feels wrong. It should feel more "hey, there's a new update!" than "you did something wrong / are using the wrong software"

I agree it's aggresive. I wanted to avoid using `"update"` because you can downgrade e.g. when latest version breaks production you can in settings set older version that should be used in studio. So it's not like "Hey there is new update" but more like "You should use different version then you're using now".

> Ignore could also be Later? Not sure if better.

I like "Later". Much better than "Ignore".

**Review by 64qam — APPROVED** [review]

**Comment by iLLiCiTiT** [comment]
I expect this notification should be modified too. This is what is shown when you hit Ignore/Later.

image

**Comment by mkolar** [comment]
@BigRoy I agree with all points with a few minor tweaks. 

**TITLE**

OpenPype update is **needed** 

_I'd use `needed` or `required` because it really shouldn't be the artist choice here, apart from possibly postponing it._

**MESSAGE**

Running OpenPype version is **3.8.0-nightly.3**. Your production has been updated to version `3.8.0-nightly.4`.


----

> Similarly Restart & Change I think should be either just **Update** or **Restart & Update**

`Restart & Update` is the best I think

> **Ignore** could also be **Later**? Not sure if better.

`Later` is great. In later PRs we can add a little dropdown with an option to postpone, so users could bypass the 5 minutes for a certain period. 

I do like the idea with a little present icon. Even though you can't really choose not to unwrap it :D

---

Regarding the notification 
**Title**: `OpenPype version is outdated`
**Message**: `Please update your OpenPype as soon as possible. To update, restart OpenPype Tray application.`

**Comment by mkolar** [comment]
>EDITED: Maybe Later could trigger the message in next interval and Ignore would stop the validation so there would be 3 buttons?

I'd leave this for another PR, but ideally it should postpone until next interval by (say 5 minutes by default), but give dropdown option with local override, same as with many notifications.
Something like this (without the `forever` option)
image

**Comment by iLLiCiTiT** [comment]
## Current state
### Dialog
image

### Notification
image

### Tray menu
- added action `Restart & Update` to tray menu which is visible when the validation do not pass
image

**Review by BigRoy — APPROVED** [review]
Haven't done a full code review nor tested locally but just going based on the screenshots. Looking good to me!

**Comment by 64qam** [comment]
The action `Restart & Update` doesn't show up in the tray menu. Everything else is ok

**Comment by iLLiCiTiT** [comment]
> The action Restart & Update doesn't show up in the tray menu. Everything else is ok

What a shame. I didn't push last commit.

**Comment by 64qam** [comment]
All good now

**Pull request merged**